Bottom: 9b231f80c679d7973d2f6badb697fa0087eb6156
Top:    984069f509ee20549bab107ad5f75b1735584cdf
Author: Ian Kent <raven@themaw.net>
Date:   2013-07-30 12:53:47 +0800

Refresh of autofs-5.0.7-add-dumpmap-to-file.patch

---

diff --git a/lib/master.c b/lib/master.c
index ec50b4c..69de543 100644
--- a/lib/master.c
+++ b/lib/master.c
@@ -1281,6 +1281,62 @@ static void list_source_instances(struct map_source *source, struct map_source *
 	return;
 }
 
+static char *match_map_name(const char *match, const char *maps)
+{
+	char *names;
+	char *map_name;
+	char *this;
+	char *tok, *ptr, *tmp;
+	unsigned int found;
+
+	tmp = strdup(match);
+	if (tmp)
+		names = strdup(maps);
+	if (!tmp || !names) {
+		if (tmp)
+			free(tmp);
+		goto fail;
+	}
+
+	if (strchr(tmp, '/'))
+		map_name = basename(tmp);
+	else
+		map_name = tmp;
+
+	found = 0;
+	this = NULL;
+	ptr = NULL;
+	tok = strtok_r(names, ",", &ptr);
+	while (tok) {
+		if (strcmp(tmp, tok)) {
+			tok = strtok_r(NULL, ",", &ptr);
+			continue;
+		}
+		found = 1;
+		this = strdup(tok);
+		break;
+	}
+
+	free(tmp);
+	free(names);
+
+	if (!this)
+		goto fail;
+
+	return this;
+
+fail:
+	if (!found)
+		printf("map \"%s\" not found in \"%s\", ignored\n",
+			match, maps);
+	else {
+		printf("failed to allocate working storage,\n");
+		printf("map %s ignored\n", match);
+	}
+
+	return NULL;
+}
+
 int master_show_mounts(struct master *master, const char *maps)
 {
 	struct list_head *p, *head;
@@ -1311,6 +1367,7 @@ int master_show_mounts(struct master *master, const char *maps)
 		struct autofs_point *ap;
 		time_t now = time(NULL);
 		unsigned int count = 0;
+		char *match_name = NULL;
 		int i;
 
 		this = list_entry(p, struct master_mapent, list);
@@ -1320,6 +1377,12 @@ int master_show_mounts(struct master *master, const char *maps)
 
 		printf("\nMount point: %s\n", ap->path);
 
+		if (maps) {
+			match_name = match_map_name(ap->path, maps);
+			if (!match_name)
+				continue;
+		}
+
 		printf("\nsource(s):\n");
 
 		/*
